app_flair.py

```diff
@@ -116,126 +126,185 @@ def parse_prediction_request(json_object: JsonDict) -> PredictionRequest:
-                    token.add_tag('ner', "E-" + ner_list[j].value)
-                    i += 1
-                    j += 1
-                else:
-                    token.add_tag('ner', 'O')
-                    i += 1
-                tokens.append(token)
-            if flag or ner_list_len == 0:
-                continue
-            while i < token_list_len:
-                token = Token(cas.get_covered_text(token_list[i]))
-                token.add_tag('ner', 'O')
-                i += 1
-                tokens.append(token)
-            s = Sentence()
-            s.tokens = tokens
-            list_sentences.append(s)
+        cas = load_cas_from_xmi(document["xmi"], typesystem=typesystem)
+        list_from_document = flair_train_ner_dataset(cas)
+        list_sentences.extend(list_from_document)
     train, dev, = train_test_split(list_sentences, train_size=0.8)
     return train, dev
 
 
-def train_ner(train, dev):
+def flair_train_ner_dataset(cas: Cas) -> list:
+    tagset = [b.decode("utf-8") for b in list(tagger_ner.tag_dictionary.idx2item)]
+    sentence_list = cas.select(SENTENCE_TYPE)
+    list_sentences = []
+    for sentence in sentence_list:
+        flag, tokens = bio_encoding(cas,sentence,tagset,Token)
+        if flag:
+            continue
+        s = Sentence()
+        s.tokens = tokens
+        list_sentences.append(s)
+    global test_sentences
+    test_sentences = list_sentences
+    # unittest.main()
+    return list_sentences
+
+
+def bio_encoding(cas,sentence,tagset,t):
+    i = 0
+    j = 0
+    token_list = list(cas.select_covered(TOKEN_TYPE, sentence))
+    ner_list = list(cas.select_covered(NER_TYPE, sentence))
+    token_list_len = len(token_list)
+    ner_list_len = len(ner_list)
+    tokens = []
+    flag = False
+    while i < token_list_len and j < ner_list_len:
+        if ner_list[j].value is None or "S-" + ner_list[j].value not in tagset:
+            flag = True
+            break
+        token = t(cas.get_covered_text(token_list[i]))
+        if (
+                token_list[i].begin == ner_list[j].begin
+                and token_list[i].end == ner_list[j].end
+        ):
+            token.add_tag("ner", "S-" + ner_list[j].value)
+            i += 1
+            j += 1
+        elif (
+                token_list[i].begin == ner_list[j].begin
+                and token_list[i].end < ner_list[j].end
+        ):
+            token.add_tag("ner", "B-" + ner_list[j].value)
+            i += 1
+        elif (
+                token_list[i].begin > ner_list[j].begin
+                and token_list[i].end < ner_list[j].end
+        ):
+            token.add_tag("ner", "I-" + ner_list[j].value)
+            i += 1
+        elif (
+                token_list[i].begin > ner_list[j].begin
+                and token_list[i].end == ner_list[j].end
+        ):
+            token.add_tag("ner", "E-" + ner_list[j].value)
```

Comment:
What is "E-"?

Reply:
"E-" is the end of a multiword ner label.